### ISSUE TITLE
Loss masking?

### ISSUE THREAD

**Issue by epii2zero** [p0p4k/pflowtts_pytorch#21]
I think loss mask m^p in paper is not implemented.

is it in todo list? or intended?

**Comment by p0p4k** [comment]
Why do you think like that?

**Comment by p0p4k** [comment]
https://github.com/p0p4k/pflowtts_pytorch/blob/297166f518263b2822f52918f86f7d27fa72fe8b/pflow/models/pflow_tts.py#L172C2-L175

**Comment by epii2zero** [comment]
I am sorry... my mistake. I misunderstood that code.
Thank you for your reply!

Can I ask some more questions?
In CFM, did you compare the performances between current implementation(Matcha style) and wavenet decoder(in paper or VITS style)

and in end2end branch,
would it make a difference to get the posterior output z_spec as normal distribution?

(In posterior encoder)
stats = self.proj(x) * x_mask      
return stats, x_mask
-->
mu = self.proj(x) * x_mask
z = (mu + torch.randn_like(mu)) * x_mask
return z, x_mask

**Comment by p0p4k** [comment]
I did not compare the difference in performance between the architectures because of lack of GPU on me right now. I expect the community to give me some feedback on that =)
E2E training is really unstable, after talking to the authors, I think doing a naturalspeech2 style - two part training is better. So you train an encodec like R-/VQ-/VAE and then use CFM to predict the VAE latents instead of mel.

**Comment by epii2zero** [comment]
oh... thank you for advice.
In my experiment, my custom model uses external duration information rather than MAS. 
Its end2end training converged.

**Comment by p0p4k** [comment]
Interesting, can you share some audio samples? Would be helpful for further development. Thanks!

**Comment by epii2zero** [comment]
How can I share?

**Comment by p0p4k** [comment]
https://vocaroo.com/?upload

**Comment by epii2zero** [comment]
My model is based on vits
and multi-speaker and multi-language model (English, Korean)
CFM is applied rather than normalizing flow.
and end2end model following your code.
posterior encoder extract latent (not Gaussian maybe)
prior loss and CFM loss target the latent.

But speech prompting is not yet applied.
it uses fixed size global condition vector.

Currently, its audio naturalness is lower than VITS(normalizing flow) based model.
after applying speech prompting, I will extend the model or postpone the end2end

English spk1 ref
https://voca.ro/1ai0cwI2ZT6f
English spk1 output
https://voca.ro/16Oqr3Gtufgi

English spk2 ref
https://voca.ro/1mNTDJrZJiC8
English spk2 output
https://voca.ro/118YqDpxio4h

Korean spk3 ref
https://voca.ro/1hfSYfImZgyO
Korean spk3 output
https://voca.ro/1ksnVqDPCrdn

**Comment by p0p4k** [comment]
Yes, it wont reach the quality unless you freeze the VITS latent (`z`). Freeze the VITS spec encoder and decoder part, let the text encoder predict `z_txt` and push it to `z` using CFM. For this use a good performing VITS pretrained base model and then transfer the corresponding weights to CFM-style-VITS model.

**Comment by epii2zero** [comment]
Okay thank you for nice discussion :)

**Comment by p0p4k** [comment]
Add me on discord (p0p4k) , 저 도 한국어 tts 관심이 많아요!

**Comment by Tera2Space** [comment]
> oh... thank you for advice. In my experiment, my custom model uses external duration information rather than MAS. Its end2end training converged.

Can you share which duration predictor do you use?

**Comment by epii2zero** [comment]
> > oh... thank you for advice. In my experiment, my custom model uses external duration information rather than MAS. Its end2end training converged.
> 
> Can you share which duration predictor do you use?

I extracted duration information using wav2vec2 trained on espeak IPA.
TTS model uses that information to align text and to train duration predictor.
